sbnanalysis/ana/SBNOsc/NumuSelection.cxx

```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
+    *ev.getValidHandle<std::vector<simb::MCTruth> >(_config.truthTag);
 
   // Iterate through the neutrinos
+  bool selected = false;
   for (size_t i=0; i<mctruths.size(); i++) {
-    Event::Interaction interaction;
     auto const& mctruth = mctruths.at(i);
+    // get the neutrino
     const simb::MCNeutrino& nu = mctruth.GetNeutrino();
 
-    if (nu.CCNC() == simb::kCC && nu.Mode() == 0 && nu.Nu().PdgCode() == 14) {
-      Event::Interaction interaction = TruthReco(mctruth);
+    // build the interaction
+    Event::Interaction interaction = TruthReco(mctruth);
+
+    // Get selection-specific info
+    //
+    // Start with the interaction stuff
+    auto intInfo = interactionInfo(ev, mctruth);
+
+    // run selection
+    std::vector<bool> selection = Select(ev, mctruth, i, intInfo);
+
+    // pass iff pass each cut
+    bool pass_selection = std::find(selection.begin(), selection.end(), false) == selection.end();
+    if (pass_selection) { 
+      // store interaction in reco tree
       reco.push_back(interaction);
+      // store local info
+      _interactionInfo->push_back(intInfo);
+
+      fNuCount++;
+      selected = true;
+    }
+    
+    // fill histos
+    unsigned select_i = 0; 
+    for (bool pass: selection) {
+      if (pass) {
+        _root_histos[select_i].h_numu_trueE->Fill(interaction.neutrino.energy);
+        _root_histos[select_i].h_numu_ccqe->Fill(ECCQE(interaction.lepton.momentum, interaction.lepton.energy));
+        _root_histos[select_i].h_numu_visibleE->Fill(intInfo.visible_energy);
+        _root_histos[select_i].h_numu_true_v_visibleE->Fill(intInfo.visible_energy - interaction.neutrino.energy);
+        _root_histos[select_i].h_numu_l_length->Fill(intInfo.l_length);
+        _root_histos[select_i].h_numu_contained_L->Fill(intInfo.l_contained_length);
+        _root_histos[select_i].h_numu_l_is_contained->Fill(intInfo.l_is_contained);
+        _root_histos[select_i].h_numu_Vxy->Fill(nu.Nu().Vx(), nu.Nu().Vy());
+        _root_histos[select_i].h_numu_Vxz->Fill(nu.Nu().Vx(), nu.Nu().Vz());
+        _root_histos[select_i].h_numu_Vyz->Fill(nu.Nu().Vy(), nu.Nu().Vz());
+      }
+      select_i ++;
+    }
+  }
+  return selected;
+}
+
+NumuSelection::NuMuInteraction NumuSelection::interactionInfo(const gallery::Event &ev, const simb::MCTruth &mctruth) {
+  // get handle to tracks and showers
+  auto const& mctrack_list = \
+    *ev.getValidHandle<std::vector<sim::MCTrack> >(_config.mctTag);
+  auto const& mcshower_list = \
+    *ev.getValidHandle<std::vector<sim::MCShower> >(_config.mcsTag);
+
+  // and particles
+  auto const& mcparticle_list = \
+    *ev.getValidHandle<std::vector<simb::MCParticle>>(_config.mcpTag);
+
+  // Get the length and determine if any point leaves the fiducial volume
+  //
+  // get lepton track
+  int lepton_ind = -1;
+  for (int i = 0; i < mctrack_list.size(); i++) {
+    if (isFromNuVertex(mctruth, mctrack_list[i]) && mctrack_list[i].PdgCode() == 13 && mctrack_list[i].size() > 0) {
```

Comment:
Also want to check here if it's a primary track?

Reply:
Fixed